rfd/0156-notifications.md

```diff
@@ -0,0 +1,423 @@
+- Product: @roraback
+
+## What
+
+A notification system that can be used to notify users of various events, updates, and warnings in real time.
+
+## Why
+
+Users currently have no way to easily view or be notified of updates and events relevant to them. Our existing cluster alerts system is cumbersome and doesn't support many
+desired use cases due to its non-dynamic nature. With the notification system, features in the application will be able to programmatically send notifications to specific user(s) to
+notify them of relevant events. Cluster admins will also be able to manually send informational notifications to their cluster's users.
+
+Examples of potential use cases include being able to notify users of:
+
+- A new access request that needs their review.
+- Their access request being approved/denied.
+- Needing to complete their periodic access list review.
+- An access monitoring alert.
+- A failed payment.
+- Approaching a usage limit.
+- Needing to upgrade their Teleport version.
+- New features added to Teleport.
+
+## Details
+
+Notifications will be stored in the backend database and be of two types:
+
+1. User-specific notifications: Targets a single specific user. An example is a notification to notify a user that their access request has been approved.
+2. Global notifications: Targets multiple users. An example is a notification to notify users that there is a new access request needing review.
+   They can be configured to target users based on one of the following matchers:
+   1. Permissions: Targets all users with the specified RBAC permission(s).
+   2. Role: Targets all users with the specified role(s).
+   3. All: Targets all users.
+
+Since user-specific notifications are stored by username, notifications for SSO users will persist between sessions.
+
+Notifications will have a default TTL of 30 days, after which they will be automatically deleted. In some cases, notifications may also automatically be deleted prior to their TTL if configured to do so.
+
+#### Backend Database
+
+User-specific notifications will be stored in the database under the key `/notifications/user/<cluster id>/<auth connector>/<the user's username>/<notification id>`.
+
+We include the `cluster id` in the key to prevent the edge case where a user from the root cluster accesses notifications in a leaf cluster that has a user with the same username, this would cause the two users' notifications
+to be merged.
+
+We include the `auth connector` in the key to prevent the edge case where a SSO user has the same username as a local user (or a user using another auth connector), this would cause the users' notifications to be merged.
+
+Global notifications will be stored in the database under the key `/notifications/global/<notification id>`.
+
+There will be a hard cap of 10,000 notifications which can exist at any given time. If a new notification is added beyond this point, it will delete and replace the oldest existing notification.
```

Comment:
Enforcing a hard cap is possible, but not without using optimistic locking of some kind (e.g. `ConditionalUpdate`, or the upcoming "transaction" api).  At the scales discussed here, that isn't really an option, as those APIs are too expensive.

I think we'd be better off rate-limiting creation so that we can't have large spikes in notifications, and then have some kind of periodic that starts actively cleaning up old notifications asynchronously once a certain size is reached.

But rate-limits and periodics are all just ways to limit the damage of a maladaptive system.  The most important thing we can do is be exceedingly careful to never create any notification that has the potential to scale linearly with cluster size (i.e. never write code that might end up worst-case generating one notification per teleport agent).